### PR TITLE
Fix 'pattern-regex' with completely empty file

### LINKED ISSUE #3705: Crash on empty file with blanket 'pattern-regex'
Using the following `pattern-regex` rule:

```yaml
rules:
- id: any-file
  message: Match any file
  languages:
    - generic
  severity: INFO
  pattern-regex: ^.*
```

With a completely empty file:

```
$ echo -n "" | docker run --rm -v /tmp/test.yaml:/tmp/test.yaml -i returntocorp/semgrep:0.61.0 --config /tmp/test.yaml -
running 1 rules...
Traceback (most recent call last):
  File "/usr/local/bin/semgrep", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.9/site-packages/semgrep/__main__.py", line 20, in main
    cli()
  File "/usr/local/lib/python3.9/site-packages/semgrep/cli.py", line 511, in cli
    semgrep.semgrep_main.main(
  File "/usr/local/lib/python3.9/site-packages/semgrep/semgrep_main.py", line 289, in main
    ignored, returned_errors = rule_match_nosem(rule_match, strict)
  File "/usr/local/lib/python3.9/site-packages/semgrep/semgrep_main.py", line 71, in rule_match_nosem
    if not rule_match.lines:
  File "/usr/local/lib/python3.9/site-packages/semgrep/rule_match.py", line 135, in lines
    result = list(itertools.islice(fd, start_line, end_line))
ValueError: Indices for islice() must be None or an integer: 0 <= x <= sys.maxsize.
```

With a 1-byte file:

```
$ echo -n "a" | docker run --rm -v /tmp/test.yaml:/tmp/test.yaml -i returntocorp/semgrep:0.61.0 --config /tmp/test.yaml -
running 1 rules...
/tmp/tmp89rr7_r3
severity:info rule:tmp.any-file: Match any file
1:a
ran 1 rules on 1 files: 1 findings
```

It looks like `rule_match` is expecting 1-indexed line numbers, but we may be receiving 0-indexed:

```python
# Start and end line are one-indexed, but the subsequent slice call is
# inclusive for start and exclusive for end, so only subtract from start
start_line = self.start["line"] - 1
end_line = self.end["line"]
```

### PR DESCRIPTION
**Pull request by mschwager**
Fixes #3705.

PR checklist:
- [x] documentation is up to date
- [x] changelog is up to date

### CODE REVIEW

File: semgrep/semgrep/rule_match.py [hunk 1]
```diff
@@ -125,6 +125,10 @@ def lines(self) -> List[str]:
         start_line = self.start["line"] - 1
         end_line = self.end["line"]
 
+        if start_line == -1 and end_line == 0:
```
Comment (nbrahms) [inline semgrep/semgrep/rule_match.py:128]: This smells off to me. Good defensive programming I guess, but semgrep shouldn't return a match here.

Comment (brendongo) [reply semgrep/semgrep/rule_match.py:128]: Should .* not match an empty file?

**Review by brendongo — APPROVED** [review]

**Comment by brendongo** [comment]
Going to kick over GithubActions

File: semgrep/semgrep/rule_match.py (hunk 1 above)
Comment (mschwager) [reply semgrep/semgrep/rule_match.py:128]: > ... but semgrep shouldn't return a match here.

IMO `pattern-regex` is a bit of an outlier, and it seems reasonable for it to return an empty match:

```
In [10]: print(re.search("^.*", ""))                                                                    
<_sre.SRE_Match object; span=(0, 0), match=''>

In [11]: print(re.search("^.+", ""))                                                                    
None
```

This type of rule is also important for our file detection rules (https://github.com/returntocorp/semgrep/issues/2438#issuecomment-891039714). In this case we don't really care if the file is empty or not, only if it exists. So I would vote that we do want to support empty matches, if only for this specific use-case.

**Pull request merged**